core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.exception.InvalidPathException;
+import alluxio.master.file.meta.MountTable;
+import alluxio.resource.CloseableResource;
+import alluxio.util.LogUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * This class is a cache from an Alluxio namespace URI ({@link AlluxioURI}, i.e. /path/to/inode) to
+ * UFS statuses.
+ *
+ * It also allows associating a path with child inodes, so that the statuses for a specific path can
+ * be searched for later.
+ */
+@ThreadSafe
+public class UfsStatusCache {
```

Comment:
I am not sure how much it would help. Do you know what the benefits are?

Much of the implementation is specialized for the usage of `InodeSyncStream`